src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -44,16 +44,41 @@ public override BoundNode VisitForEachStatement(BoundForEachStatement node)
                     return RewriteMultiDimensionalArrayForEachStatement(node);
                 }
             }
-            else if (nodeExpressionType.SpecialType == SpecialType.System_String)
+            else if (CanRewriteForEachAsFor(node.Syntax, nodeExpressionType, out var indexerGet, out var lengthGetter))
             {
-                return RewriteStringForEachStatement(node);
+                return RewriteForEachStatementAsFor(node, indexerGet, lengthGetter);
             }
             else
             {
                 return RewriteEnumeratorForEachStatement(node);
             }
         }
 
+        private bool CanRewriteForEachAsFor(SyntaxNode forEachSyntax, TypeSymbol nodeExpressionType, out MethodSymbol indexerGet, out MethodSymbol lengthGet)
+        {
+            lengthGet = indexerGet = null;
+            var origDefinition = nodeExpressionType.OriginalDefinition;
+
+            if (origDefinition.SpecialType == SpecialType.System_String)
+            {
+                lengthGet = UnsafeGetSpecialTypeMethod(forEachSyntax, SpecialMember.System_String__Length);
+                indexerGet = UnsafeGetSpecialTypeMethod(forEachSyntax, SpecialMember.System_String__Chars);
+            }
+            else if ((object)origDefinition == this._compilation.GetWellKnownType(WellKnownType.System_Span_T))
+            {
+                var spanType = (NamedTypeSymbol)nodeExpressionType;
+                lengthGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Span_T__get_Length, isOptional: true)?.SymbolAsMember(spanType);
+                indexerGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_Span_T__get_Item, isOptional: true)?.SymbolAsMember(spanType);
+            }
+            else if ((object)origDefinition == this._compilation.GetWellKnownType(WellKnownType.System_ReadOnlySpan_T))
+            {
+                var spanType = (NamedTypeSymbol)nodeExpressionType;
+                lengthGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_ReadOnlySpan_T__get_Length, isOptional: true)?.SymbolAsMember(spanType);
+                indexerGet = (MethodSymbol)_factory.WellKnownMember(WellKnownMember.System_ReadOnlySpan_T__get_Item, isOptional: true)?.SymbolAsMember(spanType);
+            }
+            return lengthGet != null && indexerGet != null;
```

Comment:
Nit: `lengthGet` and `indexerGet` are symbols. Probably need to cast to object for null check.